csharp/src/Drivers/Apache/Hive2/DecimalUtility.cs

```diff
@@ -0,0 +1,452 @@
+            bytesWritten = tempBytes.Length;
+            if (bytesWritten > bytes.Length)
+            {
+                throw new OverflowException($"Decimal size greater than {byteWidth} bytes: {bytesWritten}");
+            }
+            tempBytes.CopyTo(bytes);
+#endif
+            byte fillByte = (byte)(integerValue < 0 ? 255 : 0);
+            for (int i = bytesWritten; i < byteWidth; i++)
+            {
+                bytes[i] = fillByte;
+            }
+        }
+
+        private static BigInteger ToBigInteger(string value, int precision, int scale)
+        {
+            BigInteger intergerValue;
+#if NETCOREAPP
+            ReadOnlySpan<char> significantValue = GetSignificantValue(value, precision, scale);
+            intergerValue = BigInteger.Parse(significantValue);
+#else
+            ReadOnlySpan<char> significantValue = GetSignificantValue(value.AsSpan(), precision, scale);
+            intergerValue = BigInteger.Parse(significantValue.ToString());
+#endif
+            return intergerValue;
+        }
+
+        private static ReadOnlySpan<char> GetSignificantValue(ReadOnlySpan<char> value, int precision, int scale)
+        {
+            ParseDecimal(value, out ParserState state);
+
+            ProcessDecimal(value,
+                precision,
+                scale,
+                state,
+                out char sign,
+                out ReadOnlySpan<char> integerSpan,
+                out ReadOnlySpan<char> fractionalSpan,
+                out int neededScale);
+
+            Span<char> significant = new char[precision + 1];
+            BuildSignificantValue(
+                sign,
+                scale,
+                integerSpan,
+                fractionalSpan,
+                neededScale,
+                significant);
+
+            return significant;
+        }
+
+        private static void ProcessDecimal(ReadOnlySpan<char> value, int precision, int scale, ParserState state, out char sign, out ReadOnlySpan<char> integerSpan, out ReadOnlySpan<char> fractionalSpan, out int neededScale)
+        {
+            int int_length = 0;
+            int frac_length = 0;
+            int exponent = 0;
+
+            if (state.IntegerStart != -1 && state.IntegerEnd != -1) int_length = state.IntegerEnd - state.IntegerStart + 1;
+            if (state.FractionalStart != -1 && state.FractionalEnd != -1) frac_length = state.FractionalEnd - state.FractionalStart + 1;
+            if (state.ExponentIndex != -1 && state.ExponentStart != -1 && state.ExponentEnd != -1 && state.ExponentEnd >= state.ExponentStart)
+            {
+                int expStart = state.ExpSignIndex != -1 ? state.ExpSignIndex : state.ExponentStart;
+                int expLength = state.ExponentEnd - expStart + 1;
+                ReadOnlySpan<char> exponentSpan = value.Slice(expStart, expLength);
+#if NETCOREAPP
+                exponent = int.Parse(exponentSpan);
+#else
+                exponent = int.Parse(exponentSpan.ToString());
```

Comment:
I was thinking it was super-annoying that there are no Parse methods for `int` that take `ReadOnlySpan` in net472. Then I remembered that there are some Parse methods in System.Memory. Unfortunately, they're all for UTF-8. But then I remembered that these strings are actually stored as UTF-8 anyway and we have to pay to convert them to .NET `string`. If we would instead reference the underlying string data as UTF-8 `ReadOnlySpan`s, we can use the `TryParse` methods on `System.Buffers.Text.Utf8Parser` which include `int`, `long` and `decimal` support and for which the `decimal` implementation does allow exponents.

(Could be done as a followup.)

Reply:
@CurtHagenlocher - Sounds like a good plan. There is also parsing for `DateTime` and `DateTimeOffset`. 

In general, because all these conversions are on expected to have only ASCII characters, we may not even need to convert to string in most cases (`BigInteger.Parse` being the exception).

I'll create a follow-up task.